DataFormats/Detectors/FIT/FT0/include/DataFormatsFT0/RawEventData.h

```diff
@@ -0,0 +1,149 @@
+#include <iostream>
+#include <bitset>
+#include "Rtypes.h"
+namespace o2
+{
+namespace ft0
+{
+constexpr int NCHANNELS_FT0 = 208;
+constexpr int NCHANNELS_PM = 12;
+constexpr int NPMs = 18;
+constexpr int NBITS_EVENTDATA = 9;
+
+struct EventHeader {
+  ushort startDescriptor : 4;
+  uint reservedField : 28;
+  uint orbit : 32;
+  short bc : 12;
+  short nGBTWords : 4;
+  ClassDefNV(EventHeader, 1);
+};
+struct EventData {
+  short int time : 12;
+  short int charge : 12;
+  unsigned short int numberADC : 1;
+  bool isDoubleEvent : 1;
+  bool is1TimeLostEvent : 1;
+  bool is2TimeLostEvent : 1;
+  bool isADCinGate : 1;
+  bool isTimeInfoLate : 1;
+  bool isAmpHigh : 1;
+  bool isEventInTVDC : 1;
+  bool isTimeInfoLost : 1;
+  uint channelID : 4;
+  ClassDefNV(EventData, 1);
+};
+
+class RawEventData
+{
+ public:
+  RawEventData() = default;
+  //virtual ~RawEventData();
+  void GenerateHeader(int nChannels);
+  void GenerateData();
+  void GenerateRandomHeader(int nChannels);
+  void GenerateRandomData();
+  void GenerateRandomEvent(int nChannels);
+  EventHeader* GetEventHeaderPtr() { return &mEventHeader; }
+  EventData* GetEventDataPtr() { return mEventData; }
+  void Print(bool doPrintData = false);
+  enum EEventDataBit { kNumberADC,
+                       kIsDoubleEvent,
+                       kIs1TimeLostEvent,
+                       kIs2TimeLostEvent,
+                       kIsADCinGate,
+                       kIsTimeInfoLate,
+                       kIsAmpHigh,
+                       kIsEventInTVDC,
+                       kIsTimeInfoLost };
+  const static int gStartDescriptor = 0x0000000f;
+
+  int size() const
+  {
+    return 4                         // RAWDataHeader
+           + 1                       // EventHeader
+           + mEventHeader.nGBTWords; // EventData
+  }
+
+  void write(std::ostream& str)
+  {
+    str.write(reinterpret_cast<const char*>(&mRDH), sizeof(mRDH));
+    str.write(reinterpret_cast<const char*>(&mEventHeader), sizeof(mEventHeader));
+    char padding[16 - sizeof(mEventHeader)] = {};
+    if (mIsPadded)
+      str.write(padding, sizeof(padding));
+    //  static_assert(sizeof(mEventHeader) == 2 * sizeof(mEventData[0]), "This code assumes that pairs of data and header require the same padding");
+    for (int i = 0; i < mEventHeader.nGBTWords; ++i) {
+      str.write(reinterpret_cast<const char*>(&mEventData[2 * i]), sizeof(mEventData[0]) * 2);
+      if (mIsPadded)
+        str.write(padding, sizeof(padding));
+    }
+  }
+
+  void setIsPadded(bool isPadding128)
+  {
+    mIsPadded = isPadding128;
+  }
+
+ public:
+  o2::header::RAWDataHeader mRDH;
+  EventHeader mEventHeader;
+  EventData mEventData[NCHANNELS_PM];
+  bool mIsPadded = true;
+  /////////////////////////////////////////////////
+  ClassDefNV(RawEventData, 1);
+};
+std::ostream& operator<<(std::ostream& stream, const RawEventData& data);
+
+class DataPageWriter
+{
+  std::vector<RawEventData> mBuffer;
```

Comment:
Is mBuffer supposed to contain data of 1 trigger (1BC) or multiple ones? ``mRDH.stop`` should be set only on the last page of if a *single trigger* if its data take more than 1 CRU page. I.e. for the 1st page of the trigger the stop should be 0.
Idem, the mRDH.pageCnt should be incremented with the data of a single trigger, then it should be reset to 0.

Reply:
mBuffer contains  1 page and multiple triggers. Data from 1 BC will never be 512 words - even with full occupancy ( 208/2(max data)  + 18(GBT headers) + 18*4(RDH)) , so page will change with new orbit or after 512 words

Reply:
Each RDH opens a new page, so you are changing it for every trigger. As I wrote, you should increment the mRDH.pageCnt and set the stop=1 on last one (if >1!!!) when you are writing the data of single trigger. Alternatively, you can write multiple triggers (e.g. of the HB = orbit) in a single page under the *same* RDH, in this case your data may eventually overflow to the next page (we are now converging to this format with future RDH v5).

Reply:
But RDH has field with BC. How I will fill it for multiple BCs  under 1 RDH?

Reply:
You have your own header with the BC. In the new format to come (see https://indico.cern.ch/event/791294/contributions/3456115/attachments/1869145/3074971/wp6.pdf) the RDH v5 (see https://indico.cern.ch/event/791295/contributions/3536251/attachments/1898439/3132809/rdh.pdf) will not contain at all the trigger orbit/bc but only that of the HB. So, the 2 RDH will be used just to open and close the data of triggers collected in the HB, the identification of particular detector triggers will be done by its own headers.